### PR TITLE
Dynamic dispatch for IFunction

### LINKED ISSUE #1017: FR: dynamic dispatch of SIMD algorithms to support AVX2/AVX512
I dived into the code and addressed some SSE optimizations. Well done guys, I appreciate that you delivered and shared such a greate AP system.

And I found nothing about AVX optimizations, it is on the plan ?

### PR DESCRIPTION
**Pull request by DimasKovas**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add runtime CPU detection to select and dispatch the best function implementation. Add support for codegeneration for multiple targets. This closes #1017.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
@blinkov is a robot and he was assigned this task by mistake.

**Comment by lth08091998** [comment]
Hi, we are a small group from a company using Clickhouse in our OLAP system, and led by prof Bingsheng He( https://www.comp.nus.edu.sg/~hebs/)

We are planning to work on AVX instructions enhancement for Clickhouse too, it would be great if we can discuss with each other.

Please find me at hieu.le@bytedance.com

**Comment by alexey-milovidov** [comment]
@lth08091998 Yes, it's a good topic and this implementation of dynamic dispatch will help.
Sent you email...

File: .gitmodules [hunk 1]
```diff
@@ -157,3 +157,6 @@
 [submodule "contrib/openldap"]
 	path = contrib/openldap
 	url = https://github.com/openldap/openldap.git
+[submodule "contrib/SIMDxorshift"]
+	path = contrib/SIMDxorshift
+	url = https://github.com/lemire/SIMDxorshift
```
Comment (alexey-milovidov) [inline .gitmodules:162]: I wonder what is better - SIMD xorshift by Daniele or SIMD pcg by the same author.

File: src/Functions/TargetSpecific.h [hunk 2]
```diff
@@ -0,0 +1,176 @@
+#pragma once
+
+#include <Core/Types.h>
+
+/// This file contains macros and helpers for writing platform-dependent code.
+/// 
+/// Macroses DECLARE_<Arch>_SPECIFIC_CODE will wrap code inside them into the
```
Comment (alexey-milovidov) [inline src/Functions/TargetSpecific.h:7]: Macroses -> Macros.

File: src/Functions/TargetSpecific.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+namespace DB
+{
+
+__attribute__ ((target("xsave")))
+UInt64 xgetbv(UInt32 ecx)
+{
+    return _xgetbv(ecx);
+}
+
+UInt32 GetSupportedArches()
```
Comment (alexey-milovidov) [inline src/Functions/TargetSpecific.cpp:19]: BTW, it's already third implementation of CPU features detection :)
Previous:
- `libcpuid` from `contrib`;
- `Common/CpuId.h`

**Comment by alexey-milovidov** [comment]
Merged in #11847.

**Pull request merged**